core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,287 @@
+        this.retryPolicy = builder.retryPolicy != null ? builder.retryPolicy
+                                                       : Ec2MetadataRetryPolicy.builder().build();
+        this.endpointMode = builder.endpointMode != null ? builder.endpointMode
+                                                         : DEFAULT_ENDPOINT_PROVIDER.resolveEndpointMode();
+        this.endpoint = builder.endpoint != null ? builder.endpoint
+                                                 : URI.create(DEFAULT_ENDPOINT_PROVIDER.resolveEndpoint(this.endpointMode));
+        this.requestMarshaller = new RequestMarshaller(this.endpoint);
+        this.tokenTtl = builder.tokenTtl != null ? builder.tokenTtl
+                                                 : Duration.ofSeconds(21_600);
+        this.asyncRetryScheduler = builder.scheduledExecutorService != null
+                                   ? builder.scheduledExecutorService
+                                   : Executors.newScheduledThreadPool(DEFAULT_RETRY_THREAD_POOL_SIZE);
+        this.httpClient = builder.httpClient != null
+                          ? builder.httpClient
+                          : new DefaultSdkAsyncHttpClientBuilder().buildWithDefaults(AttributeMap.empty());
+    }
+
+    public static Ec2MetadataAsyncClient.Builder builder() {
+        return new DefaultEc2MetadataAsyncClient.Ec2MetadataAsyncBuilder();
+    }
+
+    @Override
+    public CompletableFuture<MetadataResponse> get(String path) {
+        return get(path, RetryPolicyContext.builder().retriesAttempted(0).build());
+    }
+
+    private CompletableFuture<MetadataResponse> get(String path, RetryPolicyContext retryPolicyContext) {
+        SdkHttpFullRequest baseTokenRequest = requestMarshaller.createTokenRequest(tokenTtl);
+
+        CompletableFuture<String> tokenRequest = sendAsyncRequest(baseTokenRequest);
+        CompletableFuture<MetadataResponse> result = tokenRequest
+            .thenCompose(token -> {
+                SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
+                return sendAsyncRequest(baseMetadataRequest);
+            }).thenApply(MetadataResponse::create);
+
+        return result.handle((response, error) -> {
+            if (response != null || !shouldRetry(retryPolicyContext, error)) {
+                return result;
+            }
+            int newAttempt = retryPolicyContext.retriesAttempted() + 1;
+            log.debug(() -> "Retrying request: Attempt " + newAttempt);
+            RetryPolicyContext newContext =
+                RetryPolicyContext.builder()
+                                  .request(baseTokenRequest)
+                                  .retriesAttempted(newAttempt)
+                                  .exception(SdkClientException.create(error.getMessage(), error))
+                                  .build();
+            return scheduledRetryAttempt(() -> get(path, newContext), newContext);
```

Comment:
It may be cleaner to create a return future, pass the future to `get` and completes it there than returning chained result IMO.